docs/developer/api_guidelines.md

```diff
@@ -0,0 +1,438 @@
+```
+---
+# Asynchronous Operations
+
+For requests that trigger a long running/asynchronous operation, the API endpoint **should** respond with a `202 -Accepted` HTTP response code.
+The returned payload **should** include an identifier for the job triggered by the request. This identifier should enable clients to track the current state of the
+job, or to perform other operations on it (e.g. cancelling, pausing). It **should** be suffixed with `ID`
+
+*Example:*
+
+```
+POST /projects/19340-29304:migrate
+
+Status: 202 - Accepted
+
+Body:
+
+{
+  "jobId": "123-456-789"
+}
+```
+
+After accepting and starting/queueing a long running operation, the server **may** provide the current state of the operation via a `GET` endpoint. Example:
+
+```
+GET /projects/123-456-789:migrate?job-id=123-456-789
+
+200 OK
+{
+  "status": {...}
+}
+```
+---
+# Naming Conventions
+
+## Component Names
+
+Service names **should** be concise and as simple as possible. They **should** be written in *kebap-case*. Also, since the name of a service will also be used
+for the internal domain name in K8s, they **should** adhere to the DNS label standard defined in [RFC 1035](https://datatracker.ietf.org/doc/html/rfc3339), which means that service names **should**:
+
+- Contain at most 63 characters
+- Contain only lowercase alphanumeric characters or '-'
+- Start with an alphabetic character
+- End with an alphanumeric character
+
+## Path Names
+
+Paths to an API for a specific type of resource **should** contain the resource type (in singular). E.g.:
+
+`v1/project`, and NOT `v1/projects`
+
+## Query parameters
+
+Query parameters **should** be in *kebap-case* and **should** not contain any upper case letters. For time based properties, please also refer to the [time formats](time_formats.md) section.
+
+## Field names
+
+Field names **should** be in camel case and start with a lower case character. Further, field names **should** only contain alphanumeric characters.
+Property names of JSON payloads sent to and from the API **should** always be well defined, and arbitrary property names (e.g. keys in a map) are not allowed.
+If a map should be represented, the corresponding property **should** rather be an array containing objects with properties for the key and value, respectively
+Examples:
+
+*NOT OK:*
+
+```json
+{
+  "receivedEvents": {
+    "my-arbitrary-event-type": 4,
+    "my-other-event-type": 2
+  }
+}
+```
+
+*OK:*
+
+```json
+{
+  "receivedEvents": [
+    {
+      "key": "my-arbitrary-event-type",
+      "value": 4
+    },
+    {
+      "key": "my-other-event-type",
+      "value": 2
+    }
+  ]
+}
+```
+---
+# Pagination
+
+As soon as a *listable collection* is provided via the API, it **should**
+support pagination.
+
+The endpoint
+* **must** provide a query parameter called `page-key` that points to the next page.
+  If the parameter is missing, the first page **must** be returned
+* **must** provide a query parameter called `page-size` that sets the
```

Comment:
I would also add that the endpoint must document what is the max page size it supports